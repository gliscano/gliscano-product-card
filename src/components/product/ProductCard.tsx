/* React */
import React, { createContext, CSSProperties } from 'react';
import { useProduct } from '../../hooks/useProduct';
/* interfaces */
import { OnChangeArgs, Product, ProductCardHandlers, ProductContextProps } from '../../interfaces';
/* Styles */
import styles from '../../styles/styles.module.css';
import '../../styles/custom-styles.css';
import { InitialValues } from '../../interfaces/index';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

interface ProductCardProps {
  // children?: ReactElement | ReactElement[];
  children?: (arg: ProductCardHandlers) => JSX.Element;
  className?: string;
  product: Product;
  style?: CSSProperties;
  onChange?: (arg: OnChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues
}

const ProductCard = ({
  children,
  className,
  initialValues,
  product,
  style,
  value,
  onChange,
}: ProductCardProps) => {
  const { counter, isMaxCounter, increaseBy, reset } = useProduct({ onChange, product, value, initialValues });

  return (
    <Provider value={{
      counter,
      increaseBy,
      product
    }}>
      <div
        className={`${styles.productCard} ${className}`}
        style={style}
      >
        { children && children({
            count: counter,
            isMaxCountReached: isMaxCounter,
            maxCount: initialValues?.maxCount,
            product,
            increaseBy,
            reset,
          })
        }
      </div>
    </Provider>
  );
};

export default ProductCard;