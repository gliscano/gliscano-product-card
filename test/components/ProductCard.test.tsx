import * as React from 'react';
import renderer from 'react-test-renderer'
import ProductCard from '../../src/components/product/ProductCard';
import { product1 } from '../data/products';

const { act } = renderer;

describe('Product Card', () => {
  test ('Display component', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {
          () => (
            <h4>Product Card</h4>
          )
        }
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('increase counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {
          ({ count, increaseBy }) => (
            <>
              <h4>Product Card</h4>
              <span>{ count }</span>
              <button onClick={() => increaseBy(1)}></button>
            </>
          )
        }
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();

    act(() => {
      (tree as any).children[2].props.onClick();
    });

    tree = wrapper.toJSON();
    expect((tree as any).children[1].children[0]).toBe("1");
  });
});