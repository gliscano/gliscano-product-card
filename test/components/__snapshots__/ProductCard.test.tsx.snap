// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card Display component 1`] = `
<div
  className="productCard undefined"
>
  <h4>
    Product Card
  </h4>
</div>
`;

exports[`Product Card increase counter 1`] = `
<div
  className="productCard undefined"
>
  <h4>
    Product Card
  </h4>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`Product Title Display Component with custom title 1`] = `
<span
  className="productDescription custom-class"
>
  Custom Title
</span>
`;

exports[`Product Title Display component with the product name 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - 001
  </span>
</div>
`;
